File: andriod/src/CooperHybrid/src/com/codesharp/cooper/Repositories/TaskIdxRepository.java
package com.codesharp.cooper.Repositories;

import java.util.List;

import org.json.JSONArray;
import org.json.JSONException;

import android.R.integer;
import android.R.string;
import android.content.SharedPreferences;
import android.preference.PreferenceManager;

import com.codesharp.cooper.ChangeLogDao;
import com.codesharp.cooper.Constant;
import com.codesharp.cooper.MainActivity;
import com.codesharp.cooper.Task;
import com.codesharp.cooper.TaskDao;
import com.codesharp.cooper.TaskIdx;
import com.codesharp.cooper.TaskIdxDao;
import com.codesharp.cooper.Tools;

public class TaskIdxRepository {

	private MainActivity _activity;
	private TaskIdxDao _taskIdxDao;
	private SharedPreferences _sharedPrefs;
	
	public TaskIdxRepository(MainActivity activity) {
		this._activity = activity;
		this._taskIdxDao = this._activity.getSession().getTaskIdxDao();
		this._sharedPrefs = PreferenceManager.getDefaultSharedPreferences(this._activity.getApplicationContext());
	}
	
	public List<TaskIdx> getAllTaskIdxByTemp()
	{
		List<TaskIdx> taskIdxs = this._taskIdxDao.queryBuilder()
				.where(TaskIdxDao.Properties.AccountId.eq(""))
				.list();
		return taskIdxs;
	}

	public void updateTaskIdxByNewId(String oldId, String newId, String tasklistId) throws JSONException {
		
		List<TaskIdx> taskIdxs = this.getAllTaskIdx(tasklistId);
		for (TaskIdx taskIdx : taskIdxs) {
			JSONArray sIndexesArray = new JSONArray(taskIdx.getIndexes());
			int i = 0;
			for(int index = 0; index < sIndexesArray.length(); index++)
			{
				String taskId = sIndexesArray.getString(index);
				if(taskId == oldId)
				{
					sIndexesArray.put(i, newId);
					break;
				}
				i++;
			}
			taskIdx.setIndexes(sIndexesArray.toString());
		}
		this._taskIdxDao.updateInTx(taskIdxs);
	}

	public void deleteAll(String tasklistId) {
		
		if(this._sharedPrefs.contains(Constant.USERNAME_KEY))
		{
			String username = this._sharedPrefs.getString(Constant.USERNAME_KEY, "");
			this._taskIdxDao.queryBuilder()
					.where(TaskIdxDao.Properties.AccountId.eq(username)
							, TaskIdxDao.Properties.TasklistId.eq(tasklistId))
					.buildDelete()
					.executeDeleteWithoutDetachingEntities();
		}
		else
		{
			this._taskIdxDao.queryBuilder()
					.where(TaskIdxDao.Properties.AccountId.eq("")
							, TaskIdxDao.Properties.TasklistId.eq(tasklistId))
					.buildDelete()
					.executeDeleteWithoutDetachingEntities();
		}
		//this._taskIdxDao.deleteInTx(taskIdxs);
	}

	public void updateTasklistIdByNewId(String oldId, String newId) {
		
		List<TaskIdx> taskIdxs = this.getAllTaskIdx(oldId);
		for (TaskIdx taskIdx : taskIdxs) {
			taskIdx.setTasklistId(newId);
		}
		if(taskIdxs.size() > 0)
		{
			this._taskIdxDao.updateInTx(taskIdxs);
		}
	}

	public void addTaskIdxs(List<TaskIdx> taskIdxs) {
		
		this._taskIdxDao.insertInTx(taskIdxs);
	}

	public List<TaskIdx> getAllTaskIdx(String tasklistId) {
		List<TaskIdx> taskIdxs = null;
		if(this._sharedPrefs.contains(Constant.USERNAME_KEY))
		{
			String username = this._sharedPrefs.getString(Constant.USERNAME_KEY, "");
			taskIdxs = this._taskIdxDao.queryBuilder()
					.where(TaskIdxDao.Properties.AccountId.eq(username)
							, TaskIdxDao.Properties.TasklistId.eq(tasklistId))
					.list();
		}
		else
		{
			taskIdxs = this._taskIdxDao.queryBuilder()
					.where(TaskIdxDao.Properties.AccountId.eq("")
							, TaskIdxDao.Properties.TasklistId.eq(tasklistId))
					.list();
		}
		return taskIdxs;
	}

	public void addTaskIdx(String taskId, String key, String tasklistId) {
		
		List<TaskIdx> taskIdxs = null;
		if(this._sharedPrefs.contains(Constant.USERNAME_KEY))
		{
			String username = this._sharedPrefs.getString(Constant.USERNAME_KEY, "");
			if(!Tools.isNullOrEmpty(username))
			{
				taskIdxs = this._taskIdxDao.queryBuilder()
						.where(TaskIdxDao.Properties.Key.eq(key)
								, TaskIdxDao.Properties.TasklistId.eq(tasklistId)
								, TaskIdxDao.Properties.AccountId.eq(username))
						.list();
			}
			else
			{
				taskIdxs = this._taskIdxDao.queryBuilder()
						.where(TaskIdxDao.Properties.Key.eq(key)
								, TaskIdxDao.Properties.TasklistId.eq(tasklistId)
								, TaskIdxDao.Properties.AccountId.eq(""))
						.list();
			}
		}
		
		TaskIdx taskIdx = null;
		JSONArray indexesArray = null;
		if(taskIdxs == null || (taskIdxs != null && taskIdxs.size() == 0))
		{
			taskIdx = new TaskIdx();
			taskIdx.setBy("priority");
			taskIdx.setKey(key);
			taskIdx.setName(key.equals("0") ? Constant.PRIORITY_TITLE_1 : (key.equals("1") 
					? Constant.PRIORITY_TITLE_2 : Constant.PRIORITY_TITLE_3));
			if(this._sharedPrefs.contains(Constant.USERNAME_KEY))
			{
				String username = this._sharedPrefs.getString(Constant.USERNAME_KEY, "");
				taskIdx.setAccountId(username);
			}
			else
			{
				taskIdx.setAccountId("");
			}
			indexesArray = new JSONArray();
		}
		else
		{
			taskIdx = taskIdxs.get(0);
			if(Tools.isNullOrEmpty(taskIdx.getIndexes()))
			{
				indexesArray = new JSONArray();
			}
			else
			{
				try {
					indexesArray = new JSONArray(taskIdx.getIndexes());
				} catch (JSONException e) {
					// TODO Auto-generated catch block
					e.printStackTrace();
				}
			}
		}
		indexesArray.put(taskId);
		taskIdx.setIndexes(indexesArray.toString());
		taskIdx.setTasklistId(tasklistId);
		
		if(taskIdxs == null || (taskIdxs != null && taskIdxs.size() == 0))
		{
			this._taskIdxDao.insert(taskIdx);
		}
		else 
		{
			this._taskIdxDao.update(taskIdx);
		}
	}

	public void updateTaskIdx(String taskId, String key, String tasklistId) {
		
		List<TaskIdx> taskIdxs = null;
		if(this._sharedPrefs.contains(Constant.USERNAME_KEY))
		{
			String username = this._sharedPrefs.getString(Constant.USERNAME_KEY, "");
			if(!Tools.isNullOrEmpty(username))
			{
				taskIdxs = this._taskIdxDao.queryBuilder()
						.where(TaskIdxDao.Properties.TasklistId.eq(tasklistId)
								, TaskIdxDao.Properties.AccountId.eq(username))
						.list();
			}
			else
			{
				taskIdxs = this._taskIdxDao.queryBuilder()
						.where(TaskIdxDao.Properties.TasklistId.eq(tasklistId)
								, TaskIdxDao.Properties.AccountId.eq(""))
						.list();
			}
		}
		
		if(taskIdxs != null) 
		{
			for (TaskIdx taskIdx : taskIdxs) {
				JSONArray indexesArray = null;
				JSONArray newIndexesArray = null;
				try {
					indexesArray = new JSONArray(taskIdx.getIndexes());
					newIndexesArray = new JSONArray();
					for (int i = 0; i < indexesArray.length(); i++) {
						String currentTaskId = indexesArray.getString(i);
						if(!currentTaskId.equals(taskId))
						{
							newIndexesArray.put(currentTaskId);
						}
					}
					if(taskIdx.getKey().equals(key))
					{
						newIndexesArray.put(taskId);
					}
					
				} catch (JSONException e) {
					// TODO Auto-generated catch block
					e.printStackTrace();
				}
				
				taskIdx.setIndexes(newIndexesArray.toString());
			}
			
			this._taskIdxDao.updateInTx(taskIdxs);
		}
	}

	public void deleteTaskIndexesByTaskId(String taskId, String tasklistId) {
		
		List<TaskIdx> taskIdxs = null;
		if(this._sharedPrefs.contains(Constant.USERNAME_KEY))
		{
			String username = this._sharedPrefs.getString(Constant.USERNAME_KEY, "");
			if(!Tools.isNullOrEmpty(username))
			{
				taskIdxs = this._taskIdxDao.queryBuilder()
						.where(TaskIdxDao.Properties.TasklistId.eq(tasklistId)
								, TaskIdxDao.Properties.AccountId.eq(username))
						.list();
			}
			else
			{
				taskIdxs = this._taskIdxDao.queryBuilder()
						.where(TaskIdxDao.Properties.TasklistId.eq(tasklistId)
								, TaskIdxDao.Properties.AccountId.eq(""))
						.list();
			}
		}
		
		if(taskIdxs != null) 
		{
			for (TaskIdx taskIdx : taskIdxs) {
				JSONArray indexesArray = null;
				JSONArray newIndexesArray = null;
				try {
					indexesArray = new JSONArray(taskIdx.getIndexes());
					newIndexesArray = new JSONArray();
					for (int i = 0; i < indexesArray.length(); i++) {
						String currentTaskId = indexesArray.getString(i);
						if(!currentTaskId.equals(taskId))
						{
							newIndexesArray.put(currentTaskId);
						}
					}
				} catch (JSONException e) {
					// TODO Auto-generated catch block
					e.printStackTrace();
				}
				if(newIndexesArray != null)
				{
					taskIdx.setIndexes(newIndexesArray.toString());
				}
			}
			
			this._taskIdxDao.updateInTx(taskIdxs);
		}
	}

	//批量更新任务排序
	public void updateTaskIdxs(List<TaskIdx> taskIdxs) {
		
		this._taskIdxDao.updateInTx(taskIdxs);
	}
}

File: windows phone/src/Cooper.Repositories/TaskIdxRepository.cs
using System;
using System.Collections.Generic;
using System.IO.IsolatedStorage;
using System.Linq;
using System.Net;
using System.Windows;
using System.Windows.Controls;
using System.Windows.Documents;
using System.Windows.Ink;
using System.Windows.Input;
using System.Windows.Media;
using System.Windows.Media.Animation;
using System.Windows.Shapes;
using Cooper.Core;
using Cooper.Core.Models;
using System.Collections;
using Newtonsoft.Json.Linq;

namespace Cooper.Repositories
{
    public class TaskIdxRepository
    {
        private CooperContext _context;

        public TaskIdxRepository()
        {
            this._context = new CooperContext();
        }

        public List<TaskIdx> GetAllTaskIdxByTemp()
        {
            List<TaskIdx> taskIdxs = this._context.TaskIdxs
                .Where(o => o.AccountId.Equals(""))
                .ToList();
            return taskIdxs;
        }

        public List<TaskIdx> GetAllTaskIdx(string tasklistId)
        {
            List<TaskIdx> taskIdxs = null;
            if (IsolatedStorageSettings.ApplicationSettings.Contains(Constant.USERNAME_KEY))
            {
                string username = IsolatedStorageSettings.ApplicationSettings[Constant.USERNAME_KEY] as string;
                taskIdxs = this._context.TaskIdxs
                    .Where(o => o.AccountId.Equals(username)
                        && o.TasklistId.Equals(tasklistId))
                    .ToList();
            }
            else
            {
                taskIdxs = this._context.TaskIdxs
                    .Where(o => o.AccountId.Equals("")
                        && o.TasklistId.Equals(tasklistId))
                    .ToList();
            }
            return taskIdxs;
        }

        public void UpdateTaskIdxByNewId(string oldId, string newId, string tasklistId)
        {
            List<TaskIdx> taskIdxs = this.GetAllTaskIdx(tasklistId);
            foreach (var taskIdx in taskIdxs)
            {
                JArray sIndexesArray = (JArray)taskIdx.Indexes.ToJSONObject();
                int i = 0;
                for (int index = 0; index < sIndexesArray.Count; index++)
                {
                    string taskId = sIndexesArray[index].Value<string>();
                    if (taskId == oldId)
                    {
                        sIndexesArray[i] = newId;
                        break;
                    }
                    i++;
                }
                taskIdx.Indexes = sIndexesArray.ToJSONString();
            }
            this._context.SubmitChanges();
        }

        public void DeleteAll(string tasklistId)
        {
            List<TaskIdx> taskIdxs = null;
            if (IsolatedStorageSettings.ApplicationSettings.Contains(Constant.USERNAME_KEY))
            {
                string username = IsolatedStorageSettings.ApplicationSettings[Constant.USERNAME_KEY] as string;
                taskIdxs = this._context.TaskIdxs
                    .Where(o => o.AccountId.Equals(username)
                        && o.TasklistId.Equals(tasklistId))
                        .ToList();
            }
            else
            {
                taskIdxs = this._context.TaskIdxs
                    .Where(o => o.AccountId.Equals("")
                        && o.TasklistId.Equals(tasklistId))
                        .ToList();
            }
            this._context.TaskIdxs.DeleteAllOnSubmit(taskIdxs);
            this._context.SubmitChanges();
        }

        public void UpdateTasklistIdByNewId(string oldId, string newId)
        {
            List<TaskIdx> taskIdxs = this.GetAllTaskIdx(oldId);
            foreach (var taskIdx in taskIdxs)
            {
                taskIdx.TasklistId = newId;
            }
            if (taskIdxs.Count > 0)
            {
                this._context.SubmitChanges();
            }
        }

        public void AddTaskIdxs(List<TaskIdx> taskIdxs)
        {
            this._context.TaskIdxs.InsertAllOnSubmit<TaskIdx>(taskIdxs);
            this._context.SubmitChanges();
        }

        public void AddTaskIdx(string taskId, string key, string tasklistId)
        {
            List<TaskIdx> taskIdxs = null;
            if (IsolatedStorageSettings.ApplicationSettings.Contains(Constant.USERNAME_KEY))
            {
                string username = IsolatedStorageSettings.ApplicationSettings[Constant.USERNAME_KEY] as string;
                taskIdxs = this._context.TaskIdxs
                    .Where(o => o.Key.Equals(key)
                        && o.TasklistId.Equals(tasklistId)
                        && o.AccountId.Equals(username))
                    .ToList();
            }
            else
            {
                taskIdxs = this._context.TaskIdxs
                    .Where(o => o.Key.Equals(key)
                        && o.TasklistId.Equals(tasklistId)
                        && o.AccountId.Equals(""))
                    .ToList();
            }

            TaskIdx taskIdx = null;
            JArray indexesArray = null;
            if (taskIdxs == null || (taskIdxs != null && taskIdxs.Count == 0))
            {
                taskIdx = new TaskIdx();
                taskIdx.By = "priority";
                taskIdx.Key = key;
                taskIdx.Name = key == "0" ? Constant.PRIORITY_TITLE_1 : (key == "1"
                    ? Constant.PRIORITY_TITLE_2 : Constant.PRIORITY_TITLE_3);
                if (IsolatedStorageSettings.ApplicationSettings.Contains(Constant.USERNAME_KEY))
                {
                    string username = IsolatedStorageSettings.ApplicationSettings[Constant.USERNAME_KEY] as string;
                    taskIdx.AccountId = username;
                }
                else
                {
                    taskIdx.AccountId = "";
                }
                indexesArray = new JArray();
            }
            else
            {
                taskIdx = taskIdxs[0];
                if (string.IsNullOrEmpty(taskIdx.Indexes))
                {
                    indexesArray = new JArray();
                }
                else
                {
                    indexesArray = (JArray)taskIdx.Indexes.ToJSONObject();
                }
            }
            indexesArray.Add(taskId);
            taskIdx.Indexes = indexesArray.ToJSONString();
            taskIdx.TasklistId = tasklistId;

            if (taskIdxs == null || (taskIdxs != null && taskIdxs.Count == 0))
            {
                this._context.TaskIdxs.InsertOnSubmit(taskIdx);
            }
            this._context.SubmitChanges();
        }

        public void UpdateTaskIdx(string taskId, string key, string tasklistId)
        {
            List<TaskIdx> taskIdxs = null;
            if (IsolatedStorageSettings.ApplicationSettings.Contains(Constant.USERNAME_KEY))
            {
                string username = IsolatedStorageSettings.ApplicationSettings[Constant.USERNAME_KEY] as string;
                taskIdxs = this._context.TaskIdxs
                    .Where(o => o.TasklistId.Equals(tasklistId)
                        && o.AccountId.Equals(username))
                    .ToList();
            }
            else
            {
                taskIdxs = this._context.TaskIdxs
                    .Where(o => o.TasklistId.Equals(tasklistId)
                        && o.AccountId.Equals(""))
                    .ToList();
            }

            if (taskIdxs != null)
            {
                foreach (var taskIdx in taskIdxs)
                {
                    JArray indexesArray = null;
                    JArray newIndexesArray = null;
                    indexesArray = (JArray)taskIdx.Indexes.ToJSONObject();
                    newIndexesArray = new JArray();
                    for (int i = 0; i < indexesArray.Count; i++)
                    {
                        string currentTaskId = indexesArray[i].Value<string>();
                        if (!currentTaskId.Equals(taskId))
                        {
                            newIndexesArray.Add(currentTaskId);
                        }
                    }
                    if (taskIdx.Key.Equals(key))
                    {
                        newIndexesArray.Add(taskId);
                    }
                    taskIdx.Indexes = newIndexesArray.ToJSONString();
                }

                if (taskIdxs.Count > 0)
                {
                    this._context.SubmitChanges();
                }
            }
        }

        public void DeleteTaskIndexesByTaskId(string taskId, string tasklistId)
        {
            List<TaskIdx> taskIdxs = null;
            if (IsolatedStorageSettings.ApplicationSettings.Contains(Constant.USERNAME_KEY))
            {
                string username = IsolatedStorageSettings.ApplicationSettings[Constant.USERNAME_KEY] as string;
                taskIdxs = this._context.TaskIdxs
                    .Where(o => o.TasklistId.Equals(tasklistId)
                        && o.AccountId.Equals(username))
                    .ToList();
            }
            else
            {
                taskIdxs = this._context.TaskIdxs
                    .Where(o => o.TasklistId.Equals(tasklistId)
                        && o.AccountId.Equals(""))
                    .ToList();
            }

            if (taskIdxs != null)
            {
                foreach (var taskIdx in taskIdxs)
                {
                    JArray indexesArray = null;
                    JArray newIndexesArray = null;
                    indexesArray = (JArray)taskIdx.Indexes.ToJSONObject();
                    newIndexesArray = new JArray();
                    for (int i = 0; i < indexesArray.Count; i++)
                    {
                        string currentTaskId = indexesArray[i].Value<string>();
                        if (!currentTaskId.Equals(taskId))
                        {
                            newIndexesArray.Add(currentTaskId);
                        }
                    }

                    if (newIndexesArray != null)
                    {
                        taskIdx.Indexes = newIndexesArray.ToJSONString();
                    }
                }

                if (taskIdxs.Count > 0)
                {
                    this._context.SubmitChanges();
                }
            }
        }

        public void UpdateTaskIdxs(List<TaskIdx> taskIdxs)
        {
            this._context.SubmitChanges();
        }
    }
}
